apps/examples/kernel_update_test/kernel_update_test_main.c

```diff
@@ -0,0 +1,432 @@
+		if (pre_bin_info->version != cur_bin_info->version) {
+			fail_cnt++;
+			printf("Warning! Load invalid binary.\n");
+		} else {
+			ret = OK;
+			printf("No update with invalid binary.\n");
+		}
+	}
+
+	return ret;	
+}
+
+static void binary_update_getinfo_all(void)
+{
+	int ret;
+	binary_update_info_list_t bin_info_list;
+
+	printf("\n** Binary Update GETINFO_ALL test.\n");
+	ret = binary_manager_get_update_info_all(&bin_info_list);
+	if (ret == OK) {
+		print_binary_info_list(&bin_info_list);
+	} else {
+		fail_cnt++;
+		printf("Get binary info all FAIL %d\n", ret);
+	}
+}
+
+static int binary_update_getinfo(char *name, binary_update_info_t *bin_info)
+{
+	int ret;
+
+	printf("\n** Binary Update GETINFO [%s] test.\n", name);
+	ret = binary_manager_get_update_info(name, bin_info);
+	if (ret == OK) {
+		print_binary_info(bin_info);
+	} else {
+		fail_cnt++;
+		printf("Get binary info FAIL, ret %d\n", ret);
+	}
+
+	return ret;
+}
+
+static int binary_update_reload(void)
+{
+	int ret;
+
+	printf("\n** Binary Update RELOAD test.\n");
+	ret = binary_manager_update_binary();
+	if (ret == OK) {
+		printf("RELOAD SUCCESS\n");
+	} else {
+		fail_cnt++;
+		printf("Reload binary FAIL, ret %d\n", ret);
+	}
+
+	return ret;
+}
+
+static int binary_update_same_version_test(void)
+{
+	int ret;
+	uint8_t type = 0;
+	binary_setbp_result_t result;
+	binary_update_info_t pre_bin_info;
+	binary_update_info_t cur_bin_info;
+
+	printf("\n** Binary Update Same Version Test. **\n");
+
+	ret = binary_update_getinfo(KERNEL, &pre_bin_info);
+	if (ret != OK) {
+		return ret;
+	}
+
+	/* Copy current binary with no version update */
+	ret = binary_update_download_binary(KERNEL_PATH, &pre_bin_info, false, DOWNLOAD_VALID_BIN);
+	if (ret != OK) {
+		printf("\nfailed to update same binary\n");
+		return ret;
+	}
+
+	sleep(2);
+
+	BM_SET_GROUP(type, BINARY_KERNEL);
+	ret = binary_manager_set_bootparam(type, &result);
+	if (ret != OK) {
+		if (ret == BINMGR_ALREADY_UPDATED) {
+			int idx;
+			for (idx = 0; idx < BINARY_TYPE_MAX; idx++) {
+				printf("[%d] result %d\n", idx, result.result[idx]);
+			}
+		}
+		return ret;
+	}
+
+	ret = binary_update_reload();
+	if (ret != OK) {
+		return ret;
+	}
+
+	ret = binary_update_getinfo(KERNEL, &cur_bin_info);
+	if (ret != OK) {
+		return ret;
+	}
+
+	return binary_update_check_test_result(&pre_bin_info, &cur_bin_info, DOWNLOAD_INVALID_BIN);
+}
+
+static int binary_update_new_version_test(char *bin_name)
+{
+	int ret;
+	uint8_t type = 0;
+	char path[BINARY_PATH_LEN];
+	binary_setbp_result_t result;
+	binary_update_info_t pre_bin_info;
+	binary_update_info_t cur_bin_info;
+
+	printf("\n** Binary Update New Version Test. **\n");
+
+	ret = binary_update_getinfo(bin_name, &pre_bin_info);
+	if (ret != OK) {
+		return ret;
+	}
+
+	/* Set path of binary to copy */
+	strncpy(path, KERNEL_PATH, BINARY_PATH_LEN);
+
+	/* Copy current binary and update version. */
+	ret = binary_update_download_binary(path, &pre_bin_info, true, DOWNLOAD_VALID_BIN);
```

Comment:
Could we remove duplicated code in this API & binary_update_same_version_test() ??
(only if it doesn't increase LOC's)

Reply:
Okay, will check once all the other comments are resolved.